client/common/utils.ts

```diff
@@ -539,3 +547,282 @@ export const useSetModal = () => {
 
   return { setModalVisible, selectedSet, openSetModal, closeSetModal };
 };
+
+export const getStatWithDefault = (
+  statsFromCustomSet: StatsFromCustomSet | null,
+  stat: Stat,
+) => (statsFromCustomSet ? statsFromCustomSet[stat] || 0 : 0);
+
+const getStats = (effectType: WeaponEffectType | SpellEffectType) => {
+  switch (effectType) {
+    case WeaponEffectType.AIR_DAMAGE:
+    case WeaponEffectType.AIR_STEAL:
+    case SpellEffectType.AIR_DAMAGE:
+    case SpellEffectType.AIR_STEAL:
+      return { multiplier: Stat.AGILITY, damage: Stat.AIR_DAMAGE };
+    case WeaponEffectType.EARTH_DAMAGE:
+    case WeaponEffectType.EARTH_STEAL:
+    case SpellEffectType.EARTH_DAMAGE:
+    case SpellEffectType.EARTH_STEAL:
+      return { multiplier: Stat.STRENGTH, damage: Stat.EARTH_DAMAGE };
+    case WeaponEffectType.FIRE_DAMAGE:
+    case WeaponEffectType.FIRE_STEAL:
+    case SpellEffectType.FIRE_DAMAGE:
+    case SpellEffectType.FIRE_STEAL:
+      return { multiplier: Stat.INTELLIGENCE, damage: Stat.FIRE_DAMAGE };
+    case WeaponEffectType.NEUTRAL_DAMAGE:
+    case WeaponEffectType.NEUTRAL_STEAL:
+    case SpellEffectType.NEUTRAL_DAMAGE:
+    case SpellEffectType.NEUTRAL_STEAL:
+      return { multiplier: Stat.STRENGTH, damage: Stat.NEUTRAL_DAMAGE };
+    case WeaponEffectType.WATER_DAMAGE:
+    case WeaponEffectType.WATER_STEAL:
+    case SpellEffectType.WATER_DAMAGE:
+    case SpellEffectType.WATER_STEAL:
+      return { multiplier: Stat.CHANCE, damage: Stat.WATER_DAMAGE };
+    default:
+      throw new Error('Improper effectType passed to getStats');
+  }
+};
+
+export const calcDamage = (
+  baseDamage: number,
+  effectType: WeaponEffectType | SpellEffectType,
+  stats: StatsFromCustomSet | null,
+  damageTypeInput: ICalcDamageInput,
+  weaponSkillPower?: number,
+) => {
+  const statTypes = getStats(effectType);
+  const { multiplier: multiplierType, damage: damageType } = statTypes;
+  let multiplierValue =
+    getStatWithDefault(stats, multiplierType) +
+    getStatWithDefault(stats, Stat.POWER);
+  let damageValue =
+    getStatWithDefault(stats, damageType) +
+    getStatWithDefault(stats, Stat.DAMAGE);
+  if (damageTypeInput.isTrap) {
+    multiplierValue += getStatWithDefault(stats, Stat.TRAP_POWER);
+    damageValue += getStatWithDefault(stats, Stat.TRAP_DAMAGE);
+  }
+  if (damageTypeInput.isCrit) {
+    damageValue += getStatWithDefault(stats, Stat.CRITICAL_DAMAGE);
+  }
+  if (damageTypeInput.isWeapon) {
+    multiplierValue += weaponSkillPower || 0;
+  }
+  const calculatedDamage = Math.floor(
+    baseDamage * (1 + multiplierValue / 100) + damageValue,
+  );
+  let finalDamageMod = 0;
+  if (damageTypeInput.isWeapon) {
+    finalDamageMod += getStatWithDefault(stats, Stat.PCT_WEAPON_DAMAGE);
+  } else {
+    finalDamageMod += getStatWithDefault(stats, Stat.PCT_SPELL_DAMAGE);
+  }
+
+  return {
+    melee: Math.floor(

```

Comment:
i'm almost positive that total damage bonus from final and melee/range and spell/weapon are all added up multiplicatively. something along the lines of 
`(damage_after_stats_and_flat_damage) * (1 + final damage) * (1 + melee/range damage) * (1 + spell/weapon damage)`

Reply:
dont think we've hashed out how we're going to implement final damage but that also needs to be included